### PR TITLE
Use `request_ctx` to determine whether or not `_teardown_request` should end flask span

### PR DESCRIPTION
**Pull request by matthewgrossman**
# Description

This is effectively a continuation of PR https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1654.

In that PR, we added additional context to the flask-instrumentor that ensured we wouldn't try to close spans if we weren't in the same thread that created the span. These problems are derived from `flask.copy_current_request_context`-decorated functions, which are often used when running a func in a background thread/greenlet. Because a backgrounded-function is wrapped in a copied ctx, it calls `_teardown_request` upon return, which which will fail/error when attempting to end the span.

However, problems arise even in situations where we aren't running in a separate thread/greenlet. Here's an example:

```python
import logging
import flask
from opentelemetry.instrumentation.flask import FlaskInstrumentor

app = flask.Flask(__name__)
FlaskInstrumentor.instrument_app(app=app)

@app.get("/hello")
def hello_route() -> flask.Response:

    @flask.copy_current_request_context
    def _log_header():
        print(f"inside call_on_close, {flask.request.headers.get('x-test', 'noheader')}")

    resp = flask.make_response("hello", 200)
    resp.call_on_close(_log_header)
    return resp
```
In this instance, using the `thread_id` is insufficient to differentiate in `_teardown_request` who should try to end the span; they're both running in the same thread.

Instead, we should simply keep a (weak) reference to the `request_ctx` that we're confident _did_ create the span. When a `copy_current_request_context`-func is called, it'll push on a new `request_ctx` onto the stack which won't match the original . During `_teardown_request`, we can just compare the current `request_ctx` against the original one, and escape if they don't match.

This solution should work generically no-matter how the request is run (thread/greenlet/call_on_close/synchronously). The existing multithreaded tests created in the last PR still pass (and are useful to keep around IMO)

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/868

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit test added as part of this PR, run via `tox -e 'py3{7,8,9,10,11}-test-instrumentation-flask{213,220}'`
- [x] Added new test parametrization in `tox.ini` for flask==2.1.3 and flask==2.2.0. This will exercise both conditionals of how we grab the `request_ctx`
- [x] The failing example in this PR I ran locally both on `main` and my test branch, and confirmed the issue was resolved.

<details>
  <summary>Full setup for that</summary>
  
 `app.py`:
```python
import logging
import flask
from opentelemetry.instrumentation.flask import FlaskInstrumentor

app = flask.Flask(__name__)
FlaskInstrumentor.instrument_app(app=app)

@app.get("/hello")
def hello_route() -> flask.Response:

    @flask.copy_current_request_context
    def _log_header():
        print(f"inside call_on_close, {flask.request.headers.get('x-test', 'noheader')}")

    resp = flask.make_response("hello", 200)
    resp.call_on_close(_log_header)
    return resp
```
`requirements.txt`:
```
-e ${HOME}/src/opentelemetry-python/opentelemetry-api
-e ${HOME}/src/opentelemetry-python/opentelemetry-sdk
-e ${HOME}/src/opentelemetry-python/opentelemetry-semantic-conventions
-e ${HOME}/src/opentelemetry-python-contrib/opentelemetry-instrumentation/
-e ${HOME}/src/opentelemetry-python-contrib/util/opentelemetry-util-http/
-e ${HOME}/src/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-wsgi/
-e ${HOME}/src/opentelemetry-python-contrib/instrumentation/opentelemetry-instrumentation-flask/
Flask
```

```bash
python3 -m venv venv
. venv/bin/activate
pip install -r requirements.txt
flask run
```

```bash
curl 0:5000/hello -H 'x-test: QQ'
```
</details>

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] ~Documentation has been updated~

# Implementation discussion

#### Why are you using a `weakref`?
Originally I was considering just storing the `id(request_ctx)` in the `environ`, but there's issues around potential `id` reuse [("Two objects with non-overlapping lifetimes may have the same `id()` value.")](https://docs.python.org/3/library/functions.html#id). I highly doubt there'd be strange overlapping lifetime issues that cause this to be a problem, but it just seemed better to grab a true reference to what we were interested in. This seems less brittle than effectively storing an `int` pointer to the underlying data.

The `weakref` guarantees we don't accidentally hold onto the memory simply due to this reference, as well.

#### What's with the `packaging` version checks on `Flask`?
For awhile, the blessed way of grabbing the request_ctx was `flask._request_ctx_stack.top`. This was however deprecated in version `2.2.0` ([link](https://flask.palletsprojects.com/en/2.2.x/changes/?highlight=_request_ctx_stack#version-2-2-0)). Now that we have slightly different functionality across flask versions, I also added a `flask{213,220)` section in the `tox.ini`, to ensure we test both versions

#### Why are we storing this data on the `environ` in the first place? Isn't that what `request` is for?
I'm a bit confused on this. If this file  persisted metadata using `flask.request._otel_flask_attrib`, then any new contexts spawned via `copy_current_request_context` wouldn't have inherited that data, which means teardown would do the "right thing" automatically simply by not having access to the `activation` key; itt would properly act like a stack. This _could_ be a refactor we do, but it wasn't something I wanted to do in this PR (nor am I sure if the maintainers are amenable to it).

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 1]
```diff
@@ -265,11 +266,17 @@ def response_hook(span: Span, status: str, response_headers: List):
...
 _ENVIRON_ACTIVATION_KEY = "opentelemetry-flask.activation_key"
-_ENVIRON_THREAD_ID_KEY = "opentelemetry-flask.thread_id_key"
+_ENVIRON_REQCTX_REF_KEY = "opentelemetry-flask.reqctx_ref_key"
 _ENVIRON_TOKEN = "opentelemetry-flask.token"
 
 _excluded_urls_from_env = get_excluded_urls("FLASK")
 
+if package_version.parse(flask.__version__) >= package_version.parse("2.2.0"):
+    def _request_ctx_ref() -> weakref.ReferenceType:
+        return weakref.ref(flask.globals.request_ctx._get_current_object())
```
Comment (matthewgrossman) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:276]: Used these docs to inform `_get_current_object`: https://flask.palletsprojects.com/en/2.2.x/reqcontext/#notes-on-proxies

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 2]
```diff
@@ -265,11 +266,17 @@ def response_hook(span: Span, status: str, response_headers: List):
 _ENVIRON_STARTTIME_KEY = "opentelemetry-flask.starttime_key"
 _ENVIRON_SPAN_KEY = "opentelemetry-flask.span_key"
 _ENVIRON_ACTIVATION_KEY = "opentelemetry-flask.activation_key"
-_ENVIRON_THREAD_ID_KEY = "opentelemetry-flask.thread_id_key"
+_ENVIRON_REQCTX_REF_KEY = "opentelemetry-flask.reqctx_ref_key"
 _ENVIRON_TOKEN = "opentelemetry-flask.token"
 
 _excluded_urls_from_env = get_excluded_urls("FLASK")
 
+if package_version.parse(flask.__version__) >= package_version.parse("2.2.0"):
```
Comment (matthewgrossman) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:274]: In newer versions, we're supposed to access `flask.globals.request_ctx`: https://flask.palletsprojects.com/en/2.2.x/changes/?highlight=_request_ctx_stack#version-2-2-0

File: instrumentation/opentelemetry-instrumentation-flask/pyproject.toml [hunk 3]
```diff
@@ -38,7 +39,8 @@ instruments = [
 ]
 test = [
   "opentelemetry-instrumentation-flask[instruments]",
-  "markupsafe==2.0.1",
+  "flask >= 2.2.0",
```
Comment (matthewgrossman) [inline instrumentation/opentelemetry-instrumentation-flask/pyproject.toml:42]: what would be a good way to test both pre-2.2.0 and post-2.2.0 flask in this repo?

File: dev-requirements.txt [hunk 4]
```diff
@@ -14,7 +14,7 @@ bleach==4.1.0 # transient dependency for readme-renderer
 grpcio-tools==1.29.0
 mypy-protobuf>=1.23
 protobuf~=3.13
-markupsafe==2.0.1
+markupsafe>=2.0.1
```
Comment (matthewgrossman) [inline dev-requirements.txt:17]: newer flask2.2.0 requires newer jinja2, which requires newer markupsafe

**Comment by matthewgrossman** [comment]
fyi @hangonlyra; modification/continuations of your PR

File: instrumentation/opentelemetry-instrumentation-flask/pyproject.toml (hunk 3 above)
Comment (matthewgrossman) [reply instrumentation/opentelemetry-instrumentation-flask/pyproject.toml]: figured out tox.ini changes

**Comment by hangonlyra** [comment]
> fyi @hangonlyra; modification/continuation of your PR

This is really great! Thanks for improving it.

**Comment by shalevr** [comment]
Please fix the lint errors, run "tox -e lint"

**Comment by matthewgrossman** [comment]
Thanks for the help; I got preoccupied with a bunch of things in work/life, but should come back to this PR and fix it up within the next few weeks. Will ping once its in a better state!

**Comment by matthewgrossman** [comment]
hey @shalevr; got this working locally and now would appreciate some code-review/feedback. Thanks!

**Review by ocelotl — APPROVED** [review]

**Pull request merged**